### DISCUSSION TITLE
How do I get the third-party user id after OAuth?

### DISCUSSION THREAD

**Discussion by DaniDipp** [pocketbase/pocketbase#1268] [Q&A] [answered]
How does my application get access to this id?
![2022-12-15T04-18-48](https://user-images.githubusercontent.com/17002453/207765081-864eb15a-12a4-4b20-895e-8dd77b119509.png)

I have a separate table `twitch_users` with a row for each twitch chat participant. I would like to give users who authenticate through OAuth the ability to see/update their data.

**Comment by DaniDipp** [comment]
I'd like to avoid creating a custom column on the auth collection and pass in the twitch id with the `authWithOauth2` function, because a malicious user could just put someone eles's id in there.

**Comment by ganigeorgiev** [comment, marked as answer]
As you noted, you can either pass the id to a custom field (but you'll have to use a before event hook to validate it server-side) OR you can list the user external auth providers by calling:

```js
// API:
GET /api/collections/users/records/:id/external-auths

// SDK:
const result = await pb.collection('users').listExternalAuths(pb.authStore.model.id);

// Output:
[
    {
      "id": "8171022dc95a4e8",
      "created": "2022-09-01 10:24:18.434",
      "updated": "2022-09-01 10:24:18.889",
      "recordId": "e22581b6f1d44ea",
      "collectionId": "_pb_users_auth_",
      "provider": "google",
      "providerId": "GOOGLE_USER_ID",
    },
    {
      "id": "171022dc895a4e8",
      "created": "2022-09-01 10:24:18.434",
      "updated": "2022-09-01 10:24:18.889",
      "recordId": "e22581b6f1d44ea",
      "collectionId": "_pb_users_auth_",
      "provider": "twitter",
      "providerId": "TWITTER_USER_ID",
    }
]
```

_(only admins and the user owner can access the external-auths endpoints)_

**Reply by DaniDipp** [reply]
can I also do this with filter syntax? Something like `@request.auth.externalAuths.twitch.providerId`?

**Reply by ganigeorgiev** [reply]
No, this is not supported. Could you elaborate on your use case?

**Reply by DaniDipp** [reply]
Twitch viewers are represented by virtual avatars on my stream. This data is saved in a collection `twitch_users` (`id`, `twitchId`, `body`, `hat`) and is something that my Twitch bot creates automatically for each user. I would like to give users the ability to change their hat, and thought connecting the Twitch Auth provider with the user's record in the `twitch_users` collection would be the easiest way to do this.

**Reply by ganigeorgiev** [reply]
I'm not entirely sure that I understand your use case but I've added it in my todo to eventually consider allowing access to the linked external auths from the filters.

It may not be suitable for your use case, but another possible approach would be to move the `body` and `hat` fields to the `users` collection and don't store the `twichId` anywhere.
When you want to get the linked user twich id, you can just call the `listExternalAuths` endpoint.

**Reply by DaniDipp** [reply]
Thanks, it looks like your first suggestion is actually the best way to go forward for me. Learning to extend PocketBase will certainly be useful for other stuff in the future as well.

**Reply by ganigeorgiev** [reply]
@DaniDipp You can validate and overwrite any custom client-side set value by doing something like:

```go
// unset the user defined twitchId field on external auth deletion
app.OnModelBeforeDelete().Add(func(e *core.ModelEvent) error {
	if e.Model.TableName() != "_externalAuths" {
		return nil // skip non-externalAuths
	}

	externalAuth := e.Model.(*models.ExternalAuth)
	if externalAuth.Provider != auth.NameTwitch {
		return nil // not a twitch provider
	}

	record, err := e.Dao.FindRecordById(externalAuth.CollectionId, externalAuth.RecordId)
	if err != nil {
		return apis.NewNotFoundError("The linked external auth user doesn't exist.", err)
	}

	// unset and save the record
	record.Set("twitchId", "")

	return e.Dao.SaveRecord(record)
})

// set the user defined twitchId field on external auth creation
app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
	if e.Model.TableName() != "_externalAuths" {
		return nil // skip non-users
	}

	externalAuth := e.Model.(*models.ExternalAuth)
	if externalAuth.Provider != auth.NameTwitch {
		return nil // not a twitch provider
	}

	record, err := e.Dao.FindRecordById(externalAuth.CollectionId, externalAuth.RecordId)
	if err != nil {
		return apis.NewNotFoundError("The linked external auth user doesn't exist.", err)
	}

	// set the id and save the record
	record.Set("twitchId", externalAuth.ProviderId)

	return e.Dao.SaveRecord(record)
})

// prevent client-side editing the user defined twitchId field
app.OnModelBeforeUpdate().Add(func(e *core.ModelEvent) error {
	if e.Model.TableName() != "users" {
		return nil // skip non-users
	}

	record := e.Model.(*models.Record)

	// preset the twitchId field with the available linked external provider id.
	externalAuth, _ := e.Dao.FindExternalAuthByRecordAndProvider(record,  auth.NameTwitch)
	if externalAuth != nil {
		record.Set("twitchId", externalAuth.ProviderId)
	} else {
		record.Set("twitchId", "")
	}

	// no need to explicitly call save since the finalizer of the hook will do it for us

	return nil
})
```

_Option to extend PocketBase with JS hooks will be also available in the near future._
